pkg/runtime/framework/plugins/volcano/volcano.go

```diff
@@ -0,0 +1,309 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/jobset/client-go/applyconfiguration/jobset/v1alpha2"
+	volcanov1beta1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+	volcanov1beta1ac "volcano.sh/apis/pkg/client/applyconfiguration/scheduling/v1beta1"
+
+	trainer "github.com/kubeflow/trainer/v2/pkg/apis/trainer/v1alpha1"
+	"github.com/kubeflow/trainer/v2/pkg/runtime"
+	"github.com/kubeflow/trainer/v2/pkg/runtime/framework"
+	runtimeindexer "github.com/kubeflow/trainer/v2/pkg/runtime/indexer"
+)
+
+type Volcano struct {
+	client     client.Client
+	restMapper meta.RESTMapper
+	scheme     *apiruntime.Scheme
+	logger     logr.Logger
+}
+
+var _ framework.EnforcePodGroupPolicyPlugin = (*Volcano)(nil)
+var _ framework.ComponentBuilderPlugin = (*Volcano)(nil)
+var _ framework.WatchExtensionPlugin = (*Volcano)(nil)
+
+var (
+	ErrorCanNotSetupTrainingRuntimeRuntimeClassIndexer        = errors.New("setting index on runtimeClass for TrainingRuntime")
+	ErrorCanNotSetupClusterTrainingRuntimeRuntimeClassIndexer = errors.New("setting index on runtimeClass for ClusterTrainingRuntime")
+)
+
+const Name = "Volcano"
+
+// +kubebuilder:rbac:groups=scheduling.volcano.sh,resources=podgroups,verbs=create;get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=node.k8s.io,resources=runtimeclasses,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=limitranges,verbs=get;list;watch
+
+func New(ctx context.Context, client client.Client, indexer client.FieldIndexer) (framework.Plugin, error) {
+	if err := indexer.IndexField(ctx, &trainer.TrainingRuntime{}, TrainingRuntimeContainerRuntimeClassKey,
+		IndexTrainingRuntimeContainerRuntimeClass); err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrorCanNotSetupTrainingRuntimeRuntimeClassIndexer, err)
+	}
+	if err := indexer.IndexField(ctx, &trainer.ClusterTrainingRuntime{}, ClusterTrainingRuntimeContainerRuntimeClassKey,
+		IndexClusterTrainingRuntimeContainerRuntimeClass); err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrorCanNotSetupClusterTrainingRuntimeRuntimeClassIndexer, err)
+	}
+	return &Volcano{
+		client:     client,
+		restMapper: client.RESTMapper(),
+		scheme:     client.Scheme(),
+	}, nil
+}
+
+func (v *Volcano) Name() string {
+	return Name
+}
+
+func (v *Volcano) EnforcePodGroupPolicy(info *runtime.Info, trainJob *trainer.TrainJob) error {
+	if info == nil || info.RuntimePolicy.PodGroupPolicy == nil || trainJob == nil || info.RuntimePolicy.PodGroupPolicy.Volcano == nil {
+		return nil
+	}
+	if info.Scheduler.PodLabels == nil {
+		info.Scheduler.PodLabels = map[string]string{}
+	}
+	info.Scheduler.PodLabels[volcanov1beta1.VolcanoGroupNameAnnotationKey] = trainJob.Name
+	return nil
+}
+
+func (v *Volcano) Build(ctx context.Context, info *runtime.Info, trainJob *trainer.TrainJob) ([]any, error) {
+	if info == nil || info.RuntimePolicy.PodGroupPolicy == nil || info.RuntimePolicy.PodGroupPolicy.Volcano == nil {
```

Comment:
Shall we also check if trainJob is nil?
https://github.com/kubeflow/trainer/blob/99e04625b685ceb019818d753d2cdfa048b5b11c/pkg/runtime/framework/plugins/coscheduling/coscheduling.go#L108